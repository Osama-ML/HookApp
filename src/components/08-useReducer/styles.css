body{
    background-color: rgb(95, 91, 91);
    color: rgb(0, 0, 0);
}

p{
    cursor: pointer;
    margin: 0px;
}

.list-group-item{
    display: flex;
    align-items: center;    
    justify-content: space-between;
}

.complete{
    text-decoration: line-through;
}

.container{
    padding: 50px;
}