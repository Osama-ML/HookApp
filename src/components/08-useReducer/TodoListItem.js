import React from 'react'
import './styles.css';

export const TodoListItem = ({todo, index, toggle, borrar}) => {
    return (
        <li
        key={todo.id}
        className="list-group-item"
        >
            <p 
            className={`${ todo.done && 'complete' }`}
            onClick={ () => toggle(todo.id) }
            >{index + 1}. {todo.desc}</p>
            <button 
            className="btn btn-danger"
            onClick={() => borrar(todo.id)}
            >Borrar</button>
        </li>
    )
}
