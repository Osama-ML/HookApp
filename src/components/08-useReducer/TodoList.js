import React from 'react'
import { TodoListItem } from './TodoListItem'
import './styles.css';

export const TodoList = ({todos, handleDelete, handleToggle}) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map((todo, i) => (
                    
                <TodoListItem key={todo.id} todo={todo} index={i} toggle={handleToggle} borrar={handleDelete}/>
                
                ))
            }
        </ul>
    )
}
