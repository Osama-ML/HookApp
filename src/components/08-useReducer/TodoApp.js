import React, { useEffect, useReducer } from 'react'
import { useForm } from '../../hooks/useForm';
import './styles.css';
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';



const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
} 


export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init)


    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos])

    const handleDelete = (todoId) => {

        const action = {
            type:'delete',
            payload: todoId
        }
        dispatch(action);
    }

    const handleToggle = (todoId) => {

        dispatch({
            type: 'toggle',
            payload: todoId
        })
    }

    // const handleSubmit = (e)=>{

    //     e.preventDefault();

    //     if(description.trim().length <= 1) return;

    //     const newTodo = {
    //         id: new Date().getTime(),
    //         desc: description,
    //         done: false
    //     };

    //     const action = {
    //         type:'add',
    //         payload: newTodo
    //     }
    //     dispatch(action);

    //     reset();
    // }

    const handleAddTodo = (newTodo) => {

        dispatch({
            type:'add',
            payload: newTodo
        });

    }

    return (
        <div className="container">
            <h1>TodoApp ({todos.length})</h1>
            <hr/>

            <div className="row">

                <div className="col-7">

                    <TodoList todos={todos} handleDelete={handleDelete} handleToggle={handleToggle}/>
                
                </div>

                <div className="col-5">
                    
                    <TodoAdd handleAddTodo={handleAddTodo}/>

                </div>

            </div>

        </div>
    )
}
