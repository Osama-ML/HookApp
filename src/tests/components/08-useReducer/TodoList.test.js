import { shallow } from 'enzyme';
import React from 'react';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodo } from '../../fixtures/demoTodos';

describe('pruebas en TodoList', () => {

    const handleDelete= jest.fn();
    const handleToggle= jest.fn();

    const wrapper = shallow(<TodoList 
        todos={demoTodo}
        handleDelete={handleDelete}
        handleToggle={handleToggle}
    />);

    test('debe de mostrarse correctamente', () => {
        
        expect(wrapper).toMatchSnapshot();

    })
    
    test('debe de tener dos TodoListItem', () => {
        
        expect(wrapper.find('TodoListItem').length).toBe(demoTodo.length);
        expect(wrapper.find('TodoListItem').at(0).prop('borrar')).toEqual(expect.any(Function));
    })

})
