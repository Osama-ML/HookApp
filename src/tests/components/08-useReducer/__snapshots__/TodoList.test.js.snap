// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en TodoList debe de mostrarse correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    borrar={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Learn react",
        "done": false,
        "id": 1,
      }
    }
    toggle={[MockFunction]}
  />
  <TodoListItem
    borrar={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Learn Mongo",
        "done": false,
        "id": 2,
      }
    }
    toggle={[MockFunction]}
  />
</ul>
`;
